tree.go

```diff
@@ -683,62 +685,86 @@ func (n *InternalNode) Commit() *Point {
 	internalNodeLevels := make([][]*InternalNode, StemSize)
 	n.fillLevels(internalNodeLevels)
 
-	points := make([]*Point, 0, 1024)
-	cowIndexes := make([]int, 0, 1024)
-	poly := make([]Fr, NodeWidth)
 	for level := len(internalNodeLevels) - 1; level >= 0; level-- {
 		nodes := internalNodeLevels[level]
 		if len(nodes) == 0 {
 			continue
 		}
-		points = points[:0]
-		cowIndexes = cowIndexes[:0]
-
-		// For each internal node, we collect in `points` all the ones we need to map to a field element.
-		// That is, for each touched children in a node, we collect the old and new commitment to do the diff updating
-		// later.
-		for _, node := range nodes {
-			for idx, nodeChildComm := range node.cow {
-				points = append(points, nodeChildComm)
-				points = append(points, node.children[idx].Commitment())
-				cowIndexes = append(cowIndexes, int(idx))
+
+		minBatchSize := 4
+		if len(nodes) <= minBatchSize {
+			commitNodesAtLevel(nodes)
```

Comment:
If the amount of nodes at this level is 4 or less, lets do the work directly to avoid firing goroutines and paying overhead.

Reply:
4 nodes means one inverse and a couple of microseconds per diff-commit updating, so _feels_ like a good magic number.

The reality is that this number probably depends on how fast the CPU clock is. The faster it is, the bigger `minBatchSize` might be to justify the overhead of spawning goroutines and sync them.

In the grand scheme of things, probably it isn't that important to tune it further since the bottleneck are layers that have thousands of nodes, so you'll be saturating all your cores anyway.

For example, at depth 1 most of the time we'll have to update 256 nodes since that's pretty obvious. In that case, `minBatchSize` doesn't matter much since unless you have 64 cores, your batch will have more than 4 elements. And the same applies to other layers that might have hundreds or thousands of nodes in the level, which is where most of the work is being done.

Reply:
Another angle to explain this is: if we have 16 nodes to update, and 16 cores... I don't want to spin up 16 goroutines each doing 1 thing. That's too little work per core and a lot of goroutine-sync overhead.

A more reasonable strategy is using only 4 cores of the 16 available ones, and making those 4 ones do a reasonable amount of work. That's the gist of it.